### PR TITLE
refactor(Core/Script): Refactor Void Reaver boss script to new scheduler

### PR DESCRIPTION
**Pull request by ChrisCGalbraith**
## Changes Proposed:
This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [X] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address.

## Issues Addressed:

- Refactoring to new scheduling system

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [X] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Teleport to Tempest Keep: The Eye
2. Engage Void Reaver
3. Observe abilities

## Known Issues and TODO List:

- [X] Sometimes after SPELL_POUNDING completes there are 2x SPELL_ARCANE_ORB's released. This is certainly to do with
the scheduling I've implemented.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by elthehablo** [comment]
Nice work. I am on mobile so might have missed something, but if UpdateAI only has the update victim, scheduler update and melee attack you can just omit it. It is already inherited from BossAI

**Review by elthehablo — CHANGES_REQUESTED** [review]
remove UpdateAI

**Comment by ChrisCGalbraith** [comment]
Removed the `UpdateAI` override - however there's still sometimes a double `Arcane Orb` after `Pounding` completes its channel.

**Comment by elthehablo** [comment]
> Removed the `UpdateAI` override - however there's still sometimes a double `Arcane Orb` after `Pounding` completes its channel. 

The code for it looks a bit strange. What is it supposed to do?

**Comment by elthehablo** [comment]
Generally you can do DoCastRandomTarget but range is handled a bit differently there.

**Comment by elthehablo** [comment]
Oh I know the problem. That is because scheduler queues things. If you want the orbs to consistently show up in this case you should use a different taskscheduler or you can delay events for a specific cast time. Because he's casting pounding, two orbs are queued up in the meantime

**Comment by elthehablo** [comment]
Also TaskScheduler doesn't need to be included generally. `TaskScheduler scheduler` is also inherited from BossAI. But for my above mentioned issue you might want to use this anyway. Although delaying probably works

File: src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp [hunk 1]
```diff
@@ -29,107 +30,89 @@ enum voidReaver
...
-
-        bool CheckEvadeIfOutOfCombatArea() const override
+            DoCastSelf(SPELL_BERSERK);
+        }).Schedule(15s, [this](TaskContext context)
         {
-            return me->GetDistance2d(432.59f, 371.93f) > 105.0f;
-        }
-    };
+            Talk(SAY_POUNDING);
+            DoCastSelf(SPELL_POUNDING, false);
```
Comment (elthehablo) [inline src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp:89]:
```suggestion
            DoCastSelf(SPELL_POUNDING, false);
            scheduler.DelayAll(3s);
```

File: src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp [hunk 2]
```diff
@@ -29,107 +30,89 @@ enum voidReaver
...
-            me->CallForHelp(105.0f);
-        }
+    void JustEngagedWith(Unit* who) override
+    {
+        BossAI::JustEngagedWith(who);
+        Talk(SAY_AGGRO);
+        me->CallForHelp(105.0f);
 
-        void UpdateAI(uint32 diff) override
+        scheduler.Schedule(600s, [this](TaskContext context)
```
Comment (elthehablo) [inline src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp:83]:
```suggestion
        scheduler.Schedule(10min, [this](TaskContext context)
```

very nitpicky but looks nicer

File: src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #include "ScriptMgr.h"
 #include "ScriptedCreature.h"
 #include "the_eye.h"
+#include "TaskScheduler.h"
```
Comment (elthehablo) [inline src/server/scripts/Outland/TempestKeep/Eye/boss_void_reaver.cpp:21]:
```suggestion
```
not needed if you don't use another TaskScheduler

**Review by elthehablo — CHANGES_REQUESTED** [review]

**Comment by ChrisCGalbraith** [comment]
> > Removed the `UpdateAI` override - however there's still sometimes a double `Arcane Orb` after `Pounding` completes its channel.
> 
> The code for it looks a bit strange. What is it supposed to do?

This is the original code, and I *think* that it's meant to check if there is a target further than 18 yards first, and arcane orb them. If there is no target that far away then its able to cast the orb on anyone within 20 yards. So if the whole raid is stacked they get orbed. But if you have at least one person > 18 yrds away then they can bait the orbs.

**Comment by ChrisCGalbraith** [comment]
So currently with a DelayAll(3s), the double queued arcane orbs issue is solved. However, I think it might be better to only delay the orbs if possible. Is this a situation where putting the orb schedule into a group and only delaying the group would be the better implementation?

**Comment by Nyeriah** [comment]
@ChrisCGalbraith please update your branch, the CI is stuck

**Pull request merged**